File: src/main/java/org/flossware/civilization/engine/SimulationEngine.java
package org.flossware.civilization.engine;

import org.flossware.civilization.model.*;
import org.flossware.civilization.util.SeedManager;

import java.util.*;

/**
 * Main simulation engine with variable tick loop.
 *
 * Pure function: (scenario, seed) → (finalState, eventLog)
 * Execution order (strictly sequential):
 * 1. EnvironmentModule (Climate, resources, disasters)
 * 2. MigrationModule (Population movement - placeholder for now)
 * 3. PopulationModule (Births, deaths, plague)
 * 4. EconomyModule (Production, consumption, trade)
 * 5. TechnologyModule (Research, diffusion, unlocks)
 * 6. ReligionModule (Spread, conversion, schisms)
 * 7. PoliticsModule (Stability, succession, rebellion)
 * 8. MilitaryModule (War resolution, territorial changes)
 */
public final class SimulationEngine {

    private final Scenario scenario;
    private final SeedManager seedManager;
    private final TickExecutor tickExecutor;

    public SimulationEngine(Scenario scenario, long baseSeed) {
        this.scenario = Objects.requireNonNull(scenario);
        this.seedManager = new SeedManager(baseSeed);
        TechGraph techGraph = new TechGraph(scenario.techTree());
        this.tickExecutor = new TickExecutor(scenario, techGraph);
    }

    /**
     * Runs the complete simulation from start year to end year.
     *
     * @param runIndex Monte Carlo run index for seed isolation
     * @return Final state and complete event log
     */
    public SimulationResult run(int runIndex) {
        CivilizationState state = scenario.initialState();
        List<Event> allEvents = new ArrayList<>();

        int currentYear = scenario.startYear();

        while (currentYear <= scenario.endYear()) {
            TickType tickType = TickType.determineTickType(
                scenario.worldConstraints().climateVolatility(),
                state.politics().stability()
            );

            long yearSeed = seedManager.getYearSeed(runIndex, currentYear, tickType.name());

            // Execute all modules in sequence
            TickResult tickResult = tickExecutor.executeTick(state, currentYear, tickType, yearSeed);
            state = tickResult.state();

            // Add year to all events
            final int eventYear = currentYear;
            final String civId = state.id();
            List<Event> yearEvents = tickResult.events().stream()
                .map(e -> new Event(eventYear, civId, e.type(), e.severity(), e.description(), e.data()))
                .toList();
            allEvents.addAll(yearEvents);

            currentYear += (int) Math.ceil(tickType.getYears());
        }

        return new SimulationResult(state, allEvents);
    }

    /**
     * Runs the complete simulation with periodic snapshots for charting.
     *
     * @param runIndex         Monte Carlo run index for seed isolation
     * @param snapshotInterval How many years between snapshots
     * @return Final state, event log, and periodic snapshots
     */
    public SimulationResult runWithSnapshots(int runIndex, int snapshotInterval) {
        CivilizationState state = scenario.initialState();
        List<Event> allEvents = new ArrayList<>();
        List<SimulationSnapshot> snapshots = new ArrayList<>();

        int currentYear = scenario.startYear();
        int nextSnapshot = currentYear;

        while (currentYear <= scenario.endYear()) {
            TickType tickType = TickType.determineTickType(
                scenario.worldConstraints().climateVolatility(),
                state.politics().stability()
            );

            long yearSeed = seedManager.getYearSeed(runIndex, currentYear, tickType.name());
            TickResult tickResult = tickExecutor.executeTick(state, currentYear, tickType, yearSeed);
            state = tickResult.state();

            final int eventYear = currentYear;
            final String civId = state.id();
            List<Event> yearEvents = tickResult.events().stream()
                .map(e -> new Event(eventYear, civId, e.type(), e.severity(), e.description(), e.data()))
                .toList();
            allEvents.addAll(yearEvents);

            if (currentYear >= nextSnapshot) {
                snapshots.add(new SimulationSnapshot(
                    currentYear,
                    state.population().population(),
                    state.economy().wealth(),
                    state.economy().gdp(),
                    state.technology().unlockedTechs().size(),
                    state.politics().stability(),
                    state.technology().literacyRate()
                ));
                nextSnapshot += snapshotInterval;
            }

            currentYear += (int) Math.ceil(tickType.getYears());
        }

        snapshots.add(new SimulationSnapshot(
            state.year(),
            state.population().population(),
            state.economy().wealth(),
            state.economy().gdp(),
            state.technology().unlockedTechs().size(),
            state.politics().stability(),
            state.technology().literacyRate()
        ));

        return new SimulationResult(state, allEvents, snapshots);
    }
}

File: mobile/shared/src/commonMain/kotlin/org/flossware/civilization/engine/SimulationEngine.kt
package org.flossware.civilization.engine

import org.flossware.civilization.model.*
import org.flossware.civilization.module.*
import org.flossware.civilization.util.SeedManager
import kotlin.math.ceil
import kotlin.math.max
import kotlin.math.min
import kotlin.random.Random

class SimulationEngine(
    private val scenario: Scenario,
    baseSeed: Long
) {
    private val seedManager = SeedManager(baseSeed)
    private val techGraph = TechGraph(scenario.techTree)

    fun run(runIndex: Int): SimulationResult {
        val runRandom = seedManager.getRunRandom(runIndex)
        var state = scenario.initialState
        val allEvents = mutableListOf<Event>()

        var currentYear = scenario.startYear

        while (currentYear <= scenario.endYear) {
            val tickType = TickType.determineTickType(
                scenario.worldConstraints.climateVolatility,
                state.politics.stability
            )

            val yearRandom = seedManager.getYearRandom(runRandom, currentYear, tickType.name)

            val tickResult = executeTick(state, currentYear, tickType, yearRandom)
            state = tickResult.state

            val eventYear = currentYear
            val civId = state.id
            val yearEvents = tickResult.events.map { e ->
                Event(eventYear, civId, e.type, e.severity, e.description, e.data)
            }
            allEvents.addAll(yearEvents)

            currentYear += ceil(tickType.years).toInt()
        }

        return SimulationResult(state, allEvents)
    }

    private data class TickResult(val state: CivilizationState, val events: List<Event>)

    private fun executeTick(
        initialState: CivilizationState,
        year: Int,
        tickType: TickType,
        yearRandom: Random
    ): TickResult {
        var state = initialState
        val events = mutableListOf<Event>()

        // 1. Climate Module
        val climateRandom = SeedManager.getModuleRandom(yearRandom, "climate")
        val climateResult = ClimateModule.tick(
            state.climate,
            scenario.worldConstraints.climateVolatility,
            climateRandom
        )
        state = state.withClimate(climateResult.state)
        events.addAll(climateResult.events)

        // 2. Migration Module (placeholder)

        // 3. Population Module
        val popRandom = SeedManager.getModuleRandom(yearRandom, "population")
        val popResult = PopulationModule.tick(
            state.population,
            state.climate.getResourceAbundance(),
            scenario.worldConstraints.plagueProbability,
            popRandom
        )
        state = state.withPopulation(popResult.state)
        events.addAll(popResult.events)

        // 4. Economy Module
        val econRandom = SeedManager.getModuleRandom(yearRandom, "economy")
        val econResult = EconomyModule.tick(
            state.economy,
            state.climate.getResourceAbundance(),
            state.technology.unlockedTechs,
            state.population.population,
            econRandom
        )
        state = state.withEconomy(econResult.state)
        events.addAll(econResult.events)

        // 5. Technology Module
        val techRandom = SeedManager.getModuleRandom(yearRandom, "technology")
        val techResult = TechnologyModule.tick(
            state.technology,
            techGraph,
            calculateTradeConnectivity(state),
            state.population.population,
            techRandom
        )
        state = state.withTechnology(techResult.state)
        events.addAll(techResult.events)

        // 6. Religion Module
        val relRandom = SeedManager.getModuleRandom(yearRandom, "religion")
        val relResult = ReligionModule.tick(
            state.religion,
            calculateTradeConnectivity(state),
            relRandom
        )
        state = state.withReligion(relResult.state)
        events.addAll(relResult.events)

        // 7. Politics Module
        val polRandom = SeedManager.getModuleRandom(yearRandom, "politics")
        val polResult = PoliticsModule.tick(
            state.politics,
            calculateEconomicHealth(state),
            state.religion.religiousUnity,
            state.military.atWar,
            tickType.years,
            polRandom
        )
        state = state.withPolitics(polResult.state)
        events.addAll(polResult.events)

        // 8. Military Module
        val milRandom = SeedManager.getModuleRandom(yearRandom, "military")
        val milResult = MilitaryModule.tick(
            state.military,
            state.economy.wealth,
            state.technology.unlockedTechs,
            milRandom
        )
        state = state.withMilitary(milResult.state)
        events.addAll(milResult.events)

        // 9. Random Event Module
        val randomRandom = SeedManager.getModuleRandom(yearRandom, "random_event")
        val randomResult = RandomEventModule.tick(
            state,
            scenario.worldConstraints.randomEventFrequency,
            randomRandom
        )
        state = randomResult.state
        events.addAll(randomResult.events)

        // Update year
        state = state.withYear(year)

        return TickResult(state, events)
    }

    private fun calculateTradeConnectivity(state: CivilizationState): Double {
        return min(1.0, state.economy.tradeRoutes.size * 0.2)
    }

    private fun calculateEconomicHealth(state: CivilizationState): Double {
        val wealthPerCapita = state.economy.wealth / max(1L, state.population.population)
        return min(1.0, wealthPerCapita / 1000.0)
    }
}
